modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItBuildIndexTest.java

```diff
@@ -133,23 +137,115 @@ void testDropIndexDuringTransaction() throws Exception {
 
         checkIndexBuild(partitions, replicas, INDEX_NAME);
 
+        CompletableFuture<Void> indexRemovedFuture = indexRemovedFuture();
+
         IgniteImpl node = CLUSTER.aliveNode();
 
         // Start a transaction. We expect that the index will not be removed until this transaction completes.
-        Transaction tx = node.transactions().begin(new TransactionOptions().readOnly(false));
+        node.transactions().runInTransaction(tx -> {
+            dropIndex(INDEX_NAME);
+
+            CatalogIndexDescriptor indexDescriptor = node.catalogManager().index(INDEX_NAME, node.clock().nowLong());
+
+            assertThat(indexDescriptor, is(notNullValue()));
+            assertThat(indexDescriptor.status(), is(CatalogIndexStatus.STOPPING));
+        }, new TransactionOptions().readOnly(false));
+
+        assertThat(indexRemovedFuture, willCompleteSuccessfully());
+    }
+
+    @Test
+    void testDropIndexAfterRegistering() {
+        int partitions = initialNodes();
+
+        int replicas = initialNodes();
+
+        createAndPopulateTable(replicas, partitions);
+
+        CompletableFuture<Void> indexRemovedFuture = indexRemovedFuture();
+
+        CLUSTER.aliveNode().transactions().runInTransaction(tx -> {
+            // Create an index inside a transaction, this will prevent the index from building.
+            try {
+                createIndex(INDEX_NAME);
+
+                dropIndex(INDEX_NAME);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }, new TransactionOptions().readOnly(false));
+
+        assertThat(indexRemovedFuture, willCompleteSuccessfully());
```

Comment:
Maybe check inside tx?

Reply:
This will not work, the index will not be removed until all transactions, that had observed it, have finished